### PR TITLE
improve(filesystem-hadoop): support path without scheme for gvfs api

### PR DESCRIPTION
**Pull request by coolderli**
### What changes were proposed in this pull request?

It will use the path without scheme in tensorflow. This MR will support the path without gvfs scheme.
https://github.com/tensorflow/io/blob/master/tensorflow_io/core/filesystems/hdfs/hadoop_filesystem.cc#L618
https://github.com/tensorflow/io/blob/master/tensorflow_io/core/filesystems/hdfs/hadoop_filesystem.cc#L116

### Why are the changes needed?

-  support path without Scheme for Hadoop API
- #2860

### Does this PR introduce _any_ user-facing change?

- no

### How was this patch tested?

- UTs pass

### CODE REVIEW

**Comment by coolderli** [comment]
@jerryshao @xloya Please help review this MR when you have time. Thanks.

File: clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java [hunk 1]
```diff
@@ -51,6 +53,9 @@ public class GravitinoVirtualFileSystem extends FileSystem {
   private Cache<NameIdentifier, Pair<Fileset, FileSystem>> filesetCache;
   private ScheduledThreadPoolExecutor scheduler;
 
+  private static final Pattern IDENTIFIER_PATTERN =
```
Comment (xloya) [inline clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java:56]: Could you add a description for this pattern which string will be matched?

**Comment by xloya** [comment]
Overall LGTM, left a minor suggestion.

**Comment by jerryshao** [comment]
Is it the limitation of tensorflow? If so, how does tensorflow distinguish different filesystems?

File: clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java [hunk 2]
```diff
@@ -285,27 +285,22 @@ private FileStatus convertFileStatusPathPrefix(
...
-        GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX);
-
-    if (StringUtils.isBlank(virtualUri.toString())) {
-      throw new InvalidPathException("Uri which need be extracted cannot be null or empty.");
+  @VisibleForTesting
+  NameIdentifier extractIdentifier(URI virtualUri) {
+    String virtualPath = virtualUri.toString();
+    if (StringUtils.isBlank(virtualPath)) {
+      throw new InvalidPathException(
+          virtualPath, "Uri which need be extracted cannot be null or empty.");
```
Comment (jerryshao) [inline clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java:293]: Can we use `Preconditions` for simplicity?

**Comment by jerryshao** [comment]
Seems you don't create a related issue, please please create one and associate with this PR?

File: clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java (hunk 2 above)
Comment (coolderli) [reply clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java]: done

**Comment by coolderli** [comment]
> Seems you don't create a related issue, can you please create one and associate with this PR?

create issue #2860

**Comment by coolderli** [comment]
> Is it the limitation of tensorflow? If so, how does tensorflow distinguish different filesystems?

Yes. I added a description of the issue.

Tensorflow will parseHadoopPath after creating the HadoopFileSystem(GravitinoVirtualFIleSystem). After that, it will use the path that has no scheme prefix to invoke API.

```
void ParseHadoopPath(const std::string& fname, std::string* scheme,
                     std::string* namenode, std::string* path) {
  size_t scheme_end = fname.find("://") + 2;
  // We don't want `://` in scheme.
  *scheme = fname.substr(0, scheme_end - 2);
  size_t nn_end = fname.find("/", scheme_end + 1);
  if (nn_end == std::string::npos) {
    *namenode = fname.substr(scheme_end + 1);
    *path = "";
    return;
  }
  *namenode = fname.substr(scheme_end + 1, nn_end - scheme_end - 1);
  // We keep `/` in path.
  *path = fname.substr(nn_end);  // here, truncate the scheme prefix
}
```

```
 ParseHadoopPath(path, &scheme, &namenode, &hdfs_path);

  auto handle = libhdfs->hdfsOpenFile(fs, hdfs_path.c_str(), O_WRONLY, 0, 0, 0);   // here. the hdfs_path has no gvfs:// prefix
```

File: clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java (hunk 1 above)
Comment (coolderli) [reply clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java:59]: done

**Comment by coolderli** [comment]
@jerryshao @xloya I have fixed. Please take a look. Thanks.

**Comment by coolderli** [comment]
@jerryshao Github CI is passed. Please help review this when you have time.  Thanks.

File: clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java [hunk 3]
```diff
@@ -51,6 +52,13 @@ public class GravitinoVirtualFileSystem extends FileSystem {
   private Cache<NameIdentifier, Pair<Fileset, FileSystem>> filesetCache;
   private ScheduledThreadPoolExecutor scheduler;
 
+  // The pattern is used to match gvfs path. The scheme prefix (gvfs://) is optional.
```
Comment (jerryshao) [inline clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java:55]: The scheme prefix should be "gvfs://fileset", right, not just "gvfs://"?

**Comment by jerryshao** [comment]
Just one minor issue, @coolderli can you please clarify it?

File: clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java (hunk 3 above)
Comment (coolderli) [reply clients/filesystem-hadoop3/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java]: @jerryshao You are right. I have fixed it. Please take a look. Thanks.

**Review by jerryshao — APPROVED** [review]
LGTM.

**Pull request merged**